### DISCUSSION TITLE
[4주차]_6장_키-값 저장소 설계

### DISCUSSION THREAD

**Discussion by minju412** [11th-SSAFY-19/large-scale-system-design#25] [General]
.

**Comment by minju412** [comment]
CP 시스템 설명 중, 

"은행권 시스템은 보통 데이터 일관성을 양보하지 않는다. 예를 들어, 온라인 뱅킹 시스템이 계좌 최신 정보를 출력하지 못한다면 큰 문제일 것이다" 라는 말을

"은행권 시스템은 일관성을 포기하더라도 가용성을 채택한다."는 말로 이해했는데 맞을까요?

> 96pg

**Reply by Hyunuk17** [reply]
"가용성을 포기하더라도 일관성을 채택했다"가 맞는 얘기인거 같습니다.

**Reply by minju412** [reply]
은행권 시스템을 생각해보면, 제 생각에도 가용성보다 일관성이 중요하다고 생각하는데, 
책에서 나온 "온라인 뱅킹 시스템이 계좌 최신 정보를 출력하지 못한다면 큰 문제일 것이다" 라는 말이 가용성이 중요하다라는 말로 들려서 헷갈렸습니다,,!

**Reply by minju412** [reply]
"최신 정보"를 출력하지 못하는 것 -> 오래된 정보 -> 일관성이 깨진 것

**Comment by gurwls0122** [comment]
103 그림 6-9
![image](https://github.com/user-attachments/assets/70502104-caa2-41d0-bffa-f2846d103bf5)

해당 예제가 102페이지의 충돌을 해소하기 어려운 것에 대한 방안으로 나온거 같은데.. 
전 여전히 5번 과정(충돌 해소된 데이터를 Sx가 기록)에 대한 처리 과정이 안나와 있다고 생각하는데 어떻게 생각하시나요?

요약하자면, 이는 장애 위치를 찾는 버저닝에 대한 정보만 나와있지 실제 충돌을 해소하는 방법은 잘 모르겠어서요..!

**Reply by galug** [reply]
처리하는 알고리즘 나름인 것 같습니다.

 실제로 카산드라 같은 경우에는 Lww(Last write win) 알고리즘을 사용하여 나중에 들어오는 데이터로 덮어 쓰기를 하는 방식을 취하는 것 같습니다. 
https://www.oreilly.com/library/view/learning-apache-cassandra/9781787127296/3fc4639f-9601-42a8-b4f8-d04e088891e6.xhtml

NoSQL provides a number of techniques to resolve such write conflicts. Many use LWW (last-write-wins). 아래 링크에서 보면 실제로 많은 NoSQL이 conflicts 를 해소 하기 위해서 LWW를 사용한다고 나와있습니다. 
https://dzone.com/articles/conflict-resolution-using-last-write-wins-vs-crdts
CRDT 알고리즘도 있다는데 이거는 클라이언트 단에서 충돌을 해결하는 것 같습니다. 썩 와닿지는 않네요.
https://channel.io/ko/blog/crdt_vs_ot

**Comment by gurwls0122** [comment]
104 페이지
"벡터 시계를 사용하면 어떤 버전 X가 버전 Y의 이전 버전인지(따라서 충돌이 없는지) 쉽게 판단할 수 있다.

이 얘기대로라면 모든 충돌 상황을 다 커버할 수 있는 것처럼 얘기하는 것 같은데...

위의 예시 처럼
`D5([Sx, 3], [Sy, 1], [Sz. 1])` 얘는 각각 `D3`, `D4`로 온거보다 버전이 높으니 충돌을 확인할 수 없는 것 아닐까요?

**Reply by galug** [reply]
애초에 D5 자체가 맨 앞에 [Sx, 3] 로 가장 높은 버전이므로 뒤의 버전임을 판단할 수 있습니다. 오히려 D3와 D4 가 맨 앞에 나오는 게 D5 와 같은 [Sx, 3] 일 경우에만 좀 헷갈려야하는게 맞는 것 같습니다. D3, D4 같은 경우에는 각각 [Sx, 2] 가 가장 먼저 나오기 때문에 여지가 없을 듯 합니다.

**Comment by gurwls0122** [comment]
107페이지

![image](https://github.com/user-attachments/assets/4cbd91a5-2e88-4537-9605-ac3b7a8269d0)

중재자는 어떤 것이고, 어떻게 데이터를 다른 서버로 위탁하나요..?

**Reply by galug** [reply]
중재자는 노드와 클라이언트 사이에서 proxy 역할을 한다고 나와 있습니다. 이에 대해서 별 뜻 없이 그냥 중재자 역할을 맡는 노드 하나를 중재자라고 하기도 하지만 ...
뗏목 합의 알고리즘 같은 경우는 계속해서 리더를 선출하고 이 리더가 중재자 역할을 맡게 되는 것 같습니다. 혹시 리더가 고장날 경우 후보자를 찾아서 리더를 대체하는 방법을 통해서 리더의 장애에 대한 내성을 대처하는 것 같습니다. 
https://seongjin.me/raft-consensus-algorithm/

이건 그냥 지나가다 본 블로그인데 야무져서 가져옴 
https://medium.com/rate-labs/%EC%96%B4%EB%96%A4-%EC%86%8C%EA%B0%80-%EC%9D%BC%EC%9D%84-%EB%8D%94-%EC%9E%98%ED%95%A9%EB%8B%88%EA%B9%8C-fifo%EA%B0%80-lru%EB%B3%B4%EB%8B%A4-%EB%82%AB%EC%8A%B5%EB%8B%88%EB%8B%A4%EC%9A%94-1a49b9060ce4

**Comment by gurwls0122** [comment]
108페이지 영구 장애 처리 관련해서

실제 필자가 어떤 의도로 이를 집어넣는지 모르겠는데..
애초에 가십 프로토콜과 같은 장애 감지 솔루션으로 특정 노드에 대한 장애를 감지한 순간(영구 장애)부터, 해당 노드에 데이터를 동기화 시켜줄 필요가 있나요..?
여기서 의미하는 영구장애 상태는 무엇인가요..?

그리고, 영구적인 노드 장애를 처리한다고 해놓고 예제는 한 노드에서 일관성이 망가진 데이터를 확인하는 예제를 보여주고 있는데 이를 서버 관점에서 확장이 가능한가요..?

**Reply by gurwls0122** [reply]
영구 장애, 일시 장애의 대상이 서버가 아니라 데이터의 값이 달라졌다라는 의미

**Reply by galug** [reply]
우선 영구 장애는 서버와 데이터가 영구적으로 완전히 망가졌다는 뜻이 맞습니다. 일시적인 장애는 데이터가 복원 가능하거나 서버가 복구 가능한 상태를 말하는 것도 맞습니다. 단, 머클 트리를 사용할 시 하나의 서버에서 영구적 장애가 발생한 데이터가 있을 시 서버 전 부분에 대한 전송 혹은 탐지가 필요하지 않고, 머클 트리를 통해 망가진 데이터만 O(logn) 형태로 찾아나가서 망가진 데이터를 빠르게 찾고 그 버킷만 고쳐주면 되는 것입니다.

**Comment by minju412** [comment]
`비일관성 해소 기법: 데이터 버저닝`이 낙관적 락과 유사한가요? -광민-

**Reply by galug** [reply]
위의 대답에서 힌트를 얻을 수 있을 것 같습니다. LWW 로 구현하면 먼저 들어온 게 적용되고 나중에 들어온게 적용이 안되므로 오히려 반대로 된다고 볼 수도 있을 것 같습니다.

**Comment by chwangmin** [comment]
`강한 일관성`이 낙관적락과 비슷한 방법으로 읽기/쓰기 연산을 막는 건가요?

**Reply by galug** [reply]
애초에 책을 보면 모든 사본에 현재 쓰기 연산의 결과가 반영 될 때 까지 해당 데이터에 대한 읽기/쓰기를 금지하는 것이므로 애초에 낙관적락보다는 배타락을 활용하는 것에 더 가깝다고 볼 수 있을 것 같습니다.
